provider.go

```diff
@@ -0,0 +1,599 @@
+	case DialectPostgres:
+		return postgres.New(tableName)
+	case DialectSqlite:
+		return sqlite.New(tableName)
+	}
+	return nil, fmt.Errorf("database dialect %q not yet supported", dialect)
+}
+
+func collectMigrations(fsys fs.FS, dir string) (Migrations, error) {
+	if _, err := fs.Stat(fsys, dir); errors.Is(err, fs.ErrNotExist) {
+		return nil, fmt.Errorf("%s directory does not exist", dir)
+	}
+
+	var unorderedMigrations Migrations
+
+	// Collect all SQL migration files.
+	sqlMigrationFiles, err := fs.Glob(fsys, path.Join(dir, "*.sql"))
+	if err != nil {
+		return nil, err
+	}
+	for _, fileName := range sqlMigrationFiles {
+		version, err := parseNumericComponent(fileName)
+		if err != nil {
+			return nil, err
+		}
+		unorderedMigrations = append(unorderedMigrations, &Migration{
+			Source:   fileName,
+			Version:  version,
+			Next:     -1,
+			Previous: -1,
+		})
+	}
+	// Collect Go migrations registered via goose.AddMigration().
+	for _, migration := range registeredGoMigrations {
+		if _, err := parseNumericComponent(migration.Source); err != nil {
+			return nil, fmt.Errorf("could not parse go migration file %q: %w", migration.Source, err)
+		}
+		unorderedMigrations = append(unorderedMigrations, migration)
+	}
+	// Sanity check the directory does not contain versioned Go migrations that have
+	// not been registred. This check ensures users didn't accidentally create a
+	// go migration file and forgot to register the migration.
+	//
+	// This is almost always a user error and they forgot to call: func init() { goose.AddMigration(..) }
+	if err := checkUnregisteredGoMigrations(fsys, dir); err != nil {
+		return nil, err
+	}
+	return sortAndConnectMigrations(unorderedMigrations), nil
+}
+
+func checkUnregisteredGoMigrations(fsys fs.FS, dir string) error {
+	goMigrationFiles, err := fs.Glob(fsys, path.Join(dir, "*.go"))
+	if err != nil {
+		return err
+	}
+	var unregisteredGoFiles []string
+	for _, fileName := range goMigrationFiles {
+		version, err := parseNumericComponent(fileName)
+		if err != nil {
+			// TODO(mf): log warning here?
+			// I think we do this because we allow _test.go files in the same
+			// directory.
+			continue
+		}
+		// Success, skip version because it has already been registered
+		// via goose.AddMigration().
+		if _, ok := registeredGoMigrations[version]; ok {
+			continue
+		}
+		unregisteredGoFiles = append(unregisteredGoFiles, fileName)
+	}
+	// Success, all go migration files have been registered.
+	if len(unregisteredGoFiles) == 0 {
+		return nil
+	}
+
+	f := "file"
+	if len(unregisteredGoFiles) > 1 {
+		f += "s"
+	}
+	var b strings.Builder
+
+	b.WriteString(fmt.Sprintf("error: detected %d unregistered go %s:\n", len(unregisteredGoFiles), f))
+	for _, name := range unregisteredGoFiles {
+		b.WriteString("\t" + name + "\n")
+	}
+	b.WriteString("\n")
+	b.WriteString("go functions must be registered and built into a custom binary see:\nhttps://github.com/pressly/goose/tree/master/examples/go-migrations")
```

Comment:
In the existing implementation, we did a "is registered" check in the `*Migration.run` function.

I think this is too late. Instead, we should fail much earlier when initializing the provider.

What we want to avoid is applying 3 migrations and potentially failing on the 4th if the file was not registered. 

This isn't a breaking change per se, the end result is still a failure in both cases. But doing it earlier is safer.